services/web/server/src/simcore_service_webserver/projects/_ports.py

```diff
@@ -0,0 +1,157 @@
+    """
+
+    for node_id, node in workbench.items():
+        is_inputs = filter_kind is None or filter_kind == "input"
+        is_outputs = filter_kind is None or filter_kind == "output"
+
+        # node representing INPUT ports: can write this node's output
+        if (
+            is_inputs
+            and node.key.startswith("simcore/services/frontend/parameter/")
+            and node.outputs
+        ):
+            # invariants
+            assert not node.inputs  # nosec
+            assert list(node.outputs.keys()) == ["out_1"]  # nosec
+
+            for output_key in node.outputs.keys():
+                yield _ProjectPort(
+                    kind="input", node_id=node_id, io_key=output_key, node=node
+                )
+
+        # nodes representing OUTPUT ports: can read this node's input
+        elif (
+            is_outputs
+            and node.key.startswith(
+                "simcore/services/frontend/iterator-consumer/probe/"
+            )
+            and node.inputs
+        ):
+            # invariants
+            assert not node.outputs  # nosec
+            assert list(node.inputs.keys()) == ["in_1"]  # nosec
+
+            for inputs_key in node.inputs.keys():
+                yield _ProjectPort(
+                    kind="output", node_id=node_id, io_key=inputs_key, node=node
+                )
+
+
+def get_project_inputs(workbench: dict[NodeID, Node]) -> dict[NodeID, Any]:
+    """Returns the values assigned to each input node"""
+    input_to_value = {}
+    for port in _iter_project_ports(workbench, "input"):
+        input_to_value[port.node_id] = (
+            port.node.outputs["out_1"] if port.node.outputs else None
+        )
+    return input_to_value
+
+
+class InvalidInputValue(ValueError):
+    pass
+
+
+def set_project_inputs(
+    workbench: dict[NodeID, Node], update: dict[NodeID, Any]
+) -> set[NodeID]:
+    """Updates selected input nodes and
+    Returns the ids of nodes that actually changed
+
+    raises InvalidInputValue
+    """
+    modified = set()
+    for node_id, value in update.items():
+        output = {"out_1": value}
+
+        if (node := workbench[node_id]) and node.outputs != output:
+            # validates value against jsonschema
+            try:
+                port = _ProjectPort(
+                    kind="input", node_id=node_id, io_key="out_1", node=node
+                )
+                if schema := port.get_schema():
+                    jsonschema_validate_data(value, schema)
+            except JsonSchemaValidationError as err:
+                raise InvalidInputValue(
+                    f"Invalid value for input '{node_id}': {err.message} for {value=}"
+                ) from err
+
+            workbench[node_id].outputs = output
+            modified.add(node_id)
+    return modified
+
+
+class _PortLink(PortLink):
+    class Config(PortLink.Config):
+        allow_population_by_field_name = True
```

Comment:
Maybe try to also include something hinting the difference from the original `PortLink` in the name, this only hints that the concept of PortLink is defined and used in this module.